### PR TITLE
Move mesh ghost mode into an enum

### PR DESCRIPTION
**Pull request by nate-sime**
Instead of a parameter, the mesh ghost mode is now an enum `mesh::GhostMode::(none/shared_facet/shared_vertex)`.

Added `mesh::MeshRangeType` to the python layer.

Got some more tests working.

### CODE REVIEW

File: python/test/unit/mesh/test_mesh.py [hunk 1]
```diff
@@ -364,15 +366,15 @@ def xfail_ghosted_quads_hexes(mesh_factory, ghost_mode):
...
             pytest.xfail(reason="Missing functionality in '{}' with '' "
                          "mode".format(mesh_factory, ghost_mode))
 
 
 @pytest.mark.parametrize('mesh_factory', mesh_factories_broken_shared_entities)
-def test_shared_entities(mesh_factory, ghost_mode):
+def test_shared_entities(mesh_factory):
     func, args = mesh_factory
-    xfail_ghosted_quads_hexes(func, ghost_mode)
+    # xfail_ghosted_quads_hexes(func, ghost_mode)
```
Comment (blechta) [inline python/test/unit/mesh/test_mesh.py]: Could you add ghosted meshes to `mesh_factory` above?

File: python/test/unit/fem/test_local_assembler.py [hunk 2]
```diff
@@ -46,7 +43,7 @@ def test_local_assembler_1D():
     assert near(A_matrix[1, 1], 1/60)
 
 @pytest.mark.skip
-def test_local_assembler_on_facet_integrals(ghost_mode):
+def test_local_assembler_on_facet_integrals():
     mesh = UnitSquareMesh(MPI.comm_world, 4, 4, 'right')
```
Comment (blechta) [inline python/test/unit/fem/test_local_assembler.py]: You probably want to add shared facet mode here.

File: python/test/unit/fem/test_local_assembler.py [hunk 3]
```diff
@@ -87,7 +84,7 @@ def test_local_assembler_on_facet_integrals(ghost_mode):
     assert error < 1e-8
 
 @pytest.mark.skip
-def test_local_assembler_on_facet_integrals2(ghost_mode):
+def test_local_assembler_on_facet_integrals2():
     mesh = UnitSquareMesh(MPI.comm_world, 4, 4)
```
Comment (blechta) [inline python/test/unit/fem/test_local_assembler.py]: You probably want to add shared facet mode here.

File: cpp/dolfin/mesh/Mesh.cpp [hunk 4]
```diff
@@ -395,11 +395,13 @@ std::string Mesh::str(bool verbose) const
...
 }
 //-----------------------------------------------------------------------------
-std::string Mesh::get_ghost_mode() const { return _ghost_mode; }
+mesh::GhostMode Mesh::get_ghost_mode() const
+{
+  return _ghost_mode;
+}
 //-----------------------------------------------------------------------------
-void Mesh::set_ghost_mode(std::string mode)
+void Mesh::set_ghost_mode(const GhostMode mode)
```
Comment (chrisrichardson) [inline cpp/dolfin/mesh/Mesh.cpp]: Can we remove this?

**Pull request merged**